### ISSUE TITLE
Test fms 2023.03 in UFS weather model

### ISSUE THREAD

**Issue by junwang-noaa** [ufs-community/ufs-weather-model#1981]
## Description
FMS 2023.03 release is out. The following items In this release notes may have impact on UFS WM:
```
KNOWN ISSUES:
-NO_QUAD_PRECISION macro is no longer set by FMS, the ENABLE_QUAD_PRECISION macro has replaced prior usage of NO_QUAD_PRECISION. -DENABLE_QUAD_PRECISION should be set if quad precision is to be used, otherwise FMS will not use quad precision reals where applicable.
...
Changed
MIXED PRECISION: Most subroutines and functions in FMS have been updated to simultaneously accept both 4 byte and 8 byte reals as arguments. This deprecates the --enable-mixed-mode option, which enabled similar functionality but was limited to certain directories and was not enabled by default. To facilitate easier testing of these code changes, the CMake precision options for default real size were left in (along with an equivalent --disable-r8-default flag for autotools). The resulting libraries will support mixed-precision real kinds regardless of default real size. It should also be noted that many routines that accept real arguments have been moved to include files along with headers in order to be compiled with both kinds. Most module level variables were explicitly declared as r8_kind for these updates.
```

Testing needs to be conducted to confirm FMS 2023.03 is working properly in UFS weather model.

**Comment by junwang-noaa** [comment]
@climbfuji May I ask what compile options are used when building fms library? Are "NO_QUAD_PRECISION", "--enable-mixed-mode" or "--disable-r8-default" are used when you build fms/2023.03? Thanks?

**Comment by climbfuji** [comment]
The same options as for the previous versions:
```
      variants: precision=32,64 +quad_precision +gfs_phys +openmp +pic constants=GFS build_type=Release +use_fmsio
```

**Comment by junwang-noaa** [comment]
Do you have an example of the compile line with the options? Looking at "precision=32,64 +quad_precision" ? I could not match it with in https://github.com/NOAA-GFDL/FMS/blob/2023.03/CMakeLists.txt#L59. Thanks

**Comment by climbfuji** [comment]
```
==> Installing fms-2023.03-oj2tl3talngxy77w2czu5k7c7dbjfktv
==> No binary for fms-2023.03-oj2tl3talngxy77w2czu5k7c7dbjfktv found: installing from source
==> Fetching https://github.com/NOAA-GFDL/FMS/archive/refs/tags/2023.03.tar.gz
==> No patches needed for fms
==> fms: Executing phase: 'cmake'
==> [2023-11-06-10:39:21.156930] '/opt/homebrew/bin/cmake' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:STRING=/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/fms-2023.03-oj2tl3t' '-DCMAKE_BUILD_TYPE:STRING=Release' '-DBUILD_TESTING:BOOL=OFF' '-DCMAKE_INTERPROCEDURAL_OPTIMIZATION:BOOL=OFF' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' '-DCMAKE_FIND_FRAMEWORK:STRING=LAST' '-DCMAKE_FIND_APPBUNDLE:STRING=LAST' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=ON' '-DCMAKE_INSTALL_RPATH:STRING=/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/fms-2023.03-oj2tl3t/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/fms-2023.03-oj2tl3t/lib64;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/netcdf-c-4.9.2-3kb6jx2/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/bzip2-1.0.8-vr4opg5/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/c-blosc-1.21.4-bkylaxc/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/lz4-1.9.4-rli5oxu/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/snappy-1.1.10-d4mdmgh/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/zlib-1.2.13-3rwvvpy/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/zstd-1.5.2-brsyot2/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/curl-8.1.2-j7aqucy/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/hdf5-1.14.0-he7uscp/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/libaec-1.0.6-grtvkjm/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/openmpi-4.1.5-zhlaupe/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/pmix-4.2.3-7bgnpnz/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/hwloc-2.9.1-juc4vht/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/libxml2-2.10.3-2ycuel3/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/xz-5.4.1-e44j44n/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/ncurses-6.4-n3wrrtj/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/libevent-2.1.12-pm7hwq7/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/openssl-1.1.1u-5xte3vz/lib;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/netcdf-fortran-4.6.0-qkzzwrn/lib' '-DCMAKE_PREFIX_PATH:STRING=/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/netcdf-fortran-4.6.0-qkzzwrn;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/netcdf-c-4.9.2-3kb6jx2;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/hdf5-1.14.0-he7uscp;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/openmpi-4.1.5-zhlaupe;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/pmix-4.2.3-7bgnpnz;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/libevent-2.1.12-pm7hwq7;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/openssl-1.1.1u-5xte3vz;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/hwloc-2.9.1-juc4vht;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/ncurses-6.4-n3wrrtj;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/libxml2-2.10.3-2ycuel3;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/xz-5.4.1-e44j44n;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/libaec-1.0.6-grtvkjm;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/curl-8.1.2-j7aqucy;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/c-blosc-1.21.4-bkylaxc;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/zstd-1.5.2-brsyot2;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/zlib-1.2.13-3rwvvpy;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/snappy-1.1.10-d4mdmgh;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/lz4-1.9.4-rli5oxu;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/bzip2-1.0.8-vr4opg5;/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/llvm-openmp-16.0.0-e42kivw;/opt/homebrew' '-DGFS_PHYS:BOOL=ON' '-DOPENMP:BOOL=ON' '-DENABLE_QUAD_PRECISION:BOOL=ON' '-DWITH_YAML:BOOL=OFF' '-DCONSTANTS:STRING=GFS' '-D32BIT:BOOL=ON' '-D64BIT:BOOL=ON' '-DFPIC:BOOL=ON' '-DUSE_DEPRECATED_IO:BOOL=ON' '-DCMAKE_C_COMPILER:STRING=/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/openmpi-4.1.5-zhlaupe/bin/mpicc' '-DCMAKE_CXX_COMPILER:STRING=/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/openmpi-4.1.5-zhlaupe/bin/mpic++' '-DCMAKE_Fortran_COMPILER:STRING=/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/envs/unified-env/install/apple-clang/13.1.6/openmpi-4.1.5-zhlaupe/bin/mpif90' '-DCMAKE_Fortran_FLAGS:STRING=-fallow-argument-mismatch' '/Users/heinzell/work/spack-stack/spack-stack-1.5.1-derecho/cache/build_stage/spack-stage-fms-2023.03-oj2tl3talngxy77w2czu5k7c7dbjfktv/spack-src'
```

**Comment by climbfuji** [comment]
@AlexanderRichert-NOAA @junwang-noaa We are going to tag spack-stack-1.5.1 today. I want to confirm that the cmake options for `fms@2023.03` are correct. I've condensed my reply from above into a more readable format:
```
-DCMAKE_BUILD_TYPE:STRING=Release
-DBUILD_TESTING:BOOL=OFF
-DCMAKE_INTERPROCEDURAL_OPTIMIZATION:BOOL=OFF
-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON
-DCMAKE_FIND_FRAMEWORK:STRING=LAST
-DCMAKE_FIND_APPBUNDLE:STRING=LAST
-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=ON
-DGFS_PHYS:BOOL=ON
-DOPENMP:BOOL=ON
-DENABLE_QUAD_PRECISION:BOOL=ON
-DWITH_YAML:BOOL=OFF
-DCONSTANTS:STRING=GFS
-D32BIT:BOOL=ON
-D64BIT:BOOL=ON
-DFPIC:BOOL=ON
-DUSE_DEPRECATED_IO:BOOL=ON
-DCMAKE_Fortran_FLAGS:STRING=-fallow-argument-mismatch
```

**Comment by junwang-noaa** [comment]
@climbfuji Thanks for the information. Do you have spack-stack 1.5.1 with fms 2023.03 with intel compiler? I saw your https://github.com/climbfuji/ufs-weather-model/blob/feature/hercules_gnu_gcc12mvapich237 branch has spack-stack 1.5.1 while there is no intel version of spack-stack 1.5.1. The coupled mixed mode test is only run with Intel compiler

**Comment by climbfuji** [comment]
Yes, both Intel and GNU in 1.5.1 have `fms@2023.03`. See https://spack-stack.readthedocs.io/en/1.5.1/PreConfiguredSites.html#msu-hercules  for how to load the modules. There's no branch for this (and I should delete the branch you mentioned above). spack-stack-1.5.1 on Hercules has the following versions of fms for both compilers:
```
- 2023.01
- 2023.02.01
- 2023.03
- release-jcsda
```
All other systems have everything except 2023.03 at the moment. The install can be done as soon as you are ok with the one on Hercules. Because of all the versions of fms, you can do two separate updates:
1. Update from spack-stack-1.5.0 to 1.5.1 with updates of esmf and mapl, keep fms as it is now
2. Update fms

**Comment by jokervTv** [comment]
Is it possible to build ufs now with FMS 2023.03? I found that 2023.03 is missing some symbols, but 2023.01.01 is fine.

**Comment by junwang-noaa** [comment]
@climbfuji  I am able to build and run ufs weather model with fms 2023.03 with intel on Hercules. All the gnu compile jobs failed.  

```
[junwang@hercules-login-4 modulefiles]$ git diff ufs_hercules.intel.lua
diff --git a/modulefiles/ufs_hercules.intel.lua b/modulefiles/ufs_hercules.intel.lua
index ca7e2822..605fe579 100644
--- a/modulefiles/ufs_hercules.intel.lua
+++ b/modulefiles/ufs_hercules.intel.lua
@@ -2,7 +2,7 @@ help([[
 loads UFS Model prerequisites for Hercules/Intel
 ]])

-prepend_path("MODULEPATH", "/work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.5.0/envs/unified-env/install/modulefiles/Core")
+prepend_path("MODULEPATH", "/work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.5.1/envs/unified-env/install/modulefiles/Core")

 stack_intel_ver=os.getenv("stack_intel_ver") or "2021.9.0"
 load(pathJoin("stack-intel", stack_intel_ver))
[junwang@hercules-login-4 modulefiles]$ git diff ufs_hercules.gnu.lua
diff --git a/modulefiles/ufs_hercules.gnu.lua b/modulefiles/ufs_hercules.gnu.lua
index c8ad2e49..a8c6f3c4 100644
--- a/modulefiles/ufs_hercules.gnu.lua
+++ b/modulefiles/ufs_hercules.gnu.lua
@@ -2,7 +2,7 @@ help([[
 loads UFS Model prerequisites for Hercules/GNU
 ]])

-prepend_path("MODULEPATH", "/work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.5.0/envs/unified-env-mvap2/install/modulefiles/Core")
+prepend_path("MODULEPATH", "/work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.5.1/envs/unified-env-mvap2/install/modulefiles/Core")
 -- for mvapich2, need:
 prepend_path("MODULEPATH", "/work/noaa/epic/role-epic/spack-stack/hercules/modulefiles")

```

In ufs_common.lua. I have:
```
[junwang@hercules-login-4 modulefiles]$ git diff ufs_common.lua
diff --git a/modulefiles/ufs_common.lua b/modulefiles/ufs_common.lua
index ae96ff93..3c7b714e 100644
--- a/modulefiles/ufs_common.lua
+++ b/modulefiles/ufs_common.lua
@@ -10,8 +10,8 @@ local ufs_modules = {
   {["netcdf-c"]        = "4.9.2"},
   {["netcdf-fortran"]  = "4.6.0"},
   {["parallelio"]      = "2.5.10"},
-  {["esmf"]            = "8.4.2"},
-  {["fms"]             = "2023.01"},
+  {["esmf"]            = "8.5.0"},
+  {["fms"]             = "2023.03"},
   {["bacio"]           = "2.4.1"},
   {["crtm"]            = "2.4.0"},
   {["g2"]              = "3.4.5"},
@@ -19,8 +19,8 @@ local ufs_modules = {
   {["ip"]              = "4.3.0"},
   {["sp"]              = "2.3.3"},
   {["w3emc"]           = "2.10.0"},
-  {["gftl-shared"]     = "1.5.0"},
-  {["mapl"]            = "2.35.2-esmf-8.4.2"},
+  {["gftl-shared"]     = "1.6.1"},
+  {["mapl"]            = "2.40.3-esmf-8.5.0"},
   {["scotch"]          = "7.0.4"},
 }
```


gnu compile jobs show the following error:
```
+ BUILD_DIR=/work2/noaa/stmp/junwang/stmp/junwang/FV3_RT/rt_2925923/compile_atm_gnu/build_fv3_atm_gnu
+ [[ hercules == cheyenne ]]
+ BUILD_JOBS=8
+ hostname
+ set +x
Lmod has detected the following error: These module(s) or extension(s) exist
but cannot be loaded as requested: "cmake/3.23.1", "jasper/2.0.32",
"esmf/8.5.0"
   Try: "module spider cmake/3.23.1 jasper/2.0.32 esmf/8.5.0" to see how to
load the module(s).

Executing this command requires loading "cmake/3.23.1" which failed while
processing the following module(s):

    Module fullname   Module Filename
    ---------------   ---------------
    ufs_hercules.gnu  /work/noaa/nems/junwang/ufs-weather/20231106/ufs-weather-model/modulefiles/ufs_hercules.gnu.luaExecuting this command requires loading "jasper/2.0.32" which failed while
processing the following module(s):

    Module fullname   Module Filename
    ---------------   ---------------
    ufs_common        /work/noaa/nems/junwang/ufs-weather/20231106/ufs-weather-model/modulefiles/ufs_common.lua
    ufs_hercules.gnu  /work/noaa/nems/junwang/ufs-weather/20231106/ufs-weather-model/modulefiles/ufs_hercules.gnu.luaExecuting this command requires loading "esmf/8.5.0" which failed while
processing the following module(s):

    Module fullname   Module Filename
    ---------------   ---------------
    ufs_common        /work/noaa/nems/junwang/ufs-weather/20231106/ufs-weather-model/modulefiles/ufs_common.lua
    ufs_hercules.gnu  /work/noaa/nems/junwang/ufs-weather/20231106/ufs-weather-model/modulefiles/ufs_hercules.gnu.lua

```

My code is on hecules at: /work/noaa/nems/junwang/ufs-weather/20231106/ufs-weather-model

**Comment by climbfuji** [comment]
@junwang-noaa Please refer to https://spack-stack.readthedocs.io/en/1.5.1/PreConfiguredSites.html for how to load the Hercules modules. I tested the following modifications on Hercules with GNU just now:
```
diff --git a/modulefiles/ufs_common.lua b/modulefiles/ufs_common.lua
index ae96ff93..5d184f04 100644
--- a/modulefiles/ufs_common.lua
+++ b/modulefiles/ufs_common.lua
@@ -10,8 +10,8 @@ local ufs_modules = {
   {["netcdf-c"]        = "4.9.2"},
   {["netcdf-fortran"]  = "4.6.0"},
   {["parallelio"]      = "2.5.10"},
-  {["esmf"]            = "8.4.2"},
-  {["fms"]             = "2023.01"},
+  {["esmf"]            = "8.5.0"},
+  {["fms"]             = "2023.03"},
   {["bacio"]           = "2.4.1"},
   {["crtm"]            = "2.4.0"},
   {["g2"]              = "3.4.5"},
@@ -19,8 +19,8 @@ local ufs_modules = {
   {["ip"]              = "4.3.0"},
   {["sp"]              = "2.3.3"},
   {["w3emc"]           = "2.10.0"},
-  {["gftl-shared"]     = "1.5.0"},
-  {["mapl"]            = "2.35.2-esmf-8.4.2"},
+  {["gftl-shared"]     = "1.6.2"},
+  {["mapl"]            = "2.40.3-esmf-8.5.0"},
   {["scotch"]          = "7.0.4"},
 }

diff --git a/modulefiles/ufs_hercules.gnu.lua b/modulefiles/ufs_hercules.gnu.lua
index c8ad2e49..1e4849aa 100644
--- a/modulefiles/ufs_hercules.gnu.lua
+++ b/modulefiles/ufs_hercules.gnu.lua
@@ -2,11 +2,11 @@ help([[
 loads UFS Model prerequisites for Hercules/GNU
 ]])

-prepend_path("MODULEPATH", "/work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.5.0/envs/unified-env-mvap2/install/modulefiles/Core")
+prepend_path("MODULEPATH", "/work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.5.1/envs/unified-env/install/modulefiles/Core")
 -- for mvapich2, need:
 prepend_path("MODULEPATH", "/work/noaa/epic/role-epic/spack-stack/hercules/modulefiles")

-stack_gnu_ver=os.getenv("stack_gnu_ver") or "11.3.1"
+stack_gnu_ver=os.getenv("stack_gnu_ver") or "12.2.0"
 load(pathJoin("stack-gcc", stack_gnu_ver))

 stack_mvapich2_ver=os.getenv("stack_mvapich2_ver") or "2.3.7"
```
and the following command finished successfully:
```
./compile.sh hercules "-DAPP=S2SWA -D32BIT=ON -DCCPP_SUITES=FV3_GFS_v17_coupled_p8,FV3_GFS_v17_coupled_p8_ugwpv1" "" gnu YES YES 2>&1 | tee log.compile
```

**Comment by junwang-noaa** [comment]
fms/2023.03 is tested in ufs-weather-model. It does not have the diag_table fix, which is available in fms/2023.02.01. We are going to use fms/2023.02.01 with spack-stack/1.5.1, and then move to fms/2023.04 with spack-stack/1.6.0. 

The issue is closed.
